File: Neetcode/2sum.js
//leetcode problem #1


//n runtime, n space(because of the {})
function twosum (arr, target){

    let map ={}

    for (let i=0;i<arr.length;i++){

        let targetsum = target-(arr[i])
        
        if (map[targetsum]>=0){

            return [map[targetsum], i]

        }else{
            
            map[arr[i]]=i
        }
    }

    
    return false


}

// there is another method for this Q, 
//we use nested forloops to accomlpish the same resutl,
// that solution will be n^2 time and 1 space. 

console.log(twosum([1,2,3,4,5,6], 10))

File: Neetcode/2sum.py
print ('lets learn python')

# brute force: n^2 runtime, 1 space
arr1 = [2,7,11,15]

def func(arr, target):
    for i in arr:
        j = i+1
        for j in arr:
            if (i+j==target):
                return [i,j]


print(func(arr1, 9))
   
# optimised n runtime, n space.
def twoSum(nums, target):
    prevMap={}

    for i, n in enumerate(nums):
        diff = target - n
        if diff in prevMap:
            return [prevMap[diff], i]
        prevMap[n] = i
        
    return

print(twoSum(arr1, 9))

# func()

